### ISSUE TITLE
Branch coverage symbol meanings

### ISSUE THREAD

**Issue by decimad** [gcovr/gcovr#360]
Hi there,

up till now I thought I had a mental model for what each decision's symbols mean in the html details output. I could explain most of them with "two symbols represent a conditional branch, the first symbol represents if fall-through was witnessed, the second symbol represents if branch-taken was witnessed".
However now that I had my  first look at a do .. while() construct, this rule now suddenly seems inverted, see attached screenshot where a do .. while() loop is seemingly exited by a branch-taken, even though even at machine level it is actually left by fall-through.

So my question is, is my understanding of the symbols borked, or is do .. while() an exception?

Kind regards

![gcovr capture](https://user-images.githubusercontent.com/2931955/77952135-6e23f800-72cb-11ea-88e0-46482baf84fa.png)

**Comment by latk** [comment]
TL;DR: don't make too strong assumptions about how GCC actually compiles
your code.

There is no particular logic to the meaning of individual branches, and
they should only be considered in pairs or in aggregate. The branch numbers
are provided by GCC, and depend on how GCC decides to compile the code to
machine code. You are right that the first branch tends to be fall-through,
but the choice of fallthrough branch can be surprising (unless you provide
hints via __builtin_expect(), but that isn't necessarily reliable here).

Out of interest, I've looked at one possible compilation of such a do-while
loop here https://godbolt.org/z/obFG86 (GCC 9.3 with -O0). The assembly for
your function would look similar to:

  mov edx, 0
.LOOP:
  add edx, 1
  cmp edx, eax  ; compare counter <=> stop
  jge .END  ; if counter >= stop: goto END
  jmp .LOOP
.END:
  mov eax, edx
  ret

Thus, the loop termination is not compiled as a fallthrough under -O0. Of
course this changes at higher optimization levels, where you need to
introduce a side effect into the loop to prevent it from being optimized
away entirely (see the Godbolt link and set to -O3).

On the html-rewrite branch of gcovr, we're doing away with these check
marks and provide only a summary like "1/2", with further details hidden in
a popup. Unfortunately I don't have the time/energy right now to get it
ready for a release, but hopefully soonishly.

**Comment by decimad** [comment]
Hello there, first off - thanks for your elaborate answer, it surely wasn't TL;DR;
Now I have to ask for pardon because I actually doublechecked, and yes my GCC version also generates branch-out-of-loop code with the inverted loop invariant, I just didn't read carefully :(. With "-O0" that is of course. So it actually IS consistent with my former understanding, just that the implementation is a bit un-intuitive (to me).
Also I double-checked: with optimizations it is transformed into fall-through-out-of-loop with positive loop invariant, yet the symbols stay the same.
This makes me believe that the compiler actually tries to give the counters the initially mentioned semantics when generating the "basic code", but optimizing transformations happen afterwards without transforming the counter semantics.

But I'm curious, is it not somehow a "common" problem to try to reason which condition (conditional branch) failed to achieve coverage and why, given the coverage results?

**Comment by Spacetown** [comment]
Can this issue be closed? As mentioned by @latk is the detailed report changed.
